iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -4,1310 +4,1110 @@
-        public override bool IsUsable => State != TransportState.Closed && State != TransportState.Error;
-        public TimeSpan TwinTimeout { get; set; } = s_defaultTwinTimeout;
+        private bool CertificateValidationHandler(MqttClientCertificateValidationEventArgs args)
+        {
+            return _mqttTransportSettings.RemoteCertificateValidationCallback.Invoke(
+                _mqttClient,
+                args.Certificate,
+                args.Chain,
+                args.SslPolicyErrors);
+        }
 
         #region Client operations
 
         public override async Task OpenAsync(CancellationToken cancellationToken)
         {
-            try
+            string clientId = string.IsNullOrWhiteSpace(_moduleId) ? _deviceId : $"{_deviceId}/{_moduleId}";
+            _mqttClientOptionsBuilder.WithClientId(clientId);
+
+            string username = $"{_hostName}/{clientId}/?{ClientApiVersionHelper.ApiVersionQueryStringLatest}&DeviceClientType={Uri.EscapeDataString(_productInfo.ToString())}";
+
+            if (!string.IsNullOrWhiteSpace(_modelId))
             {
-                if (Logging.IsEnabled)
-                    Logging.Enter(this, cancellationToken, nameof(OpenAsync));
+                username += $"&model-id={Uri.EscapeDataString(_modelId)}";
+            }
+
+            if (!string.IsNullOrWhiteSpace(_mqttTransportSettings?.AuthenticationChain))
+            {
+                username += $"&auth-chain={Uri.EscapeDataString(_mqttTransportSettings.AuthenticationChain)}";
+            }
 
-                cancellationToken.ThrowIfCancellationRequested();
+            if (_connectionCredentials.SasTokenRefresher != null)
+            {
+                // Symmetric key authenticated connections need to set client Id, username, and password
+                string password = await _connectionCredentials.SasTokenRefresher.GetTokenAsync(_connectionCredentials.IotHubHostName).ConfigureAwait(false);
+                _mqttClientOptionsBuilder.WithCredentials(username, password);
+            }
+            else if (_connectionCredentials.SharedAccessSignature != null)
+            {
+                // Symmetric key authenticated connections need to set client Id, username, and password
+                string password = _connectionCredentials.SharedAccessSignature;
+                _mqttClientOptionsBuilder.WithCredentials(username, password);
+            }
+            else
+            {
+                // x509 authenticated connections only need to set client Id and username
+                _mqttClientOptionsBuilder.WithCredentials(username);
+            }
 
-                EnsureValidState(throwIfNotOpen: false);
+            _mqttClientOptions = _mqttClientOptionsBuilder.Build();
 
-                await OpenInternalAsync(cancellationToken).ConfigureAwait(false);
+            try
+            {
+                _mqttClient.DisconnectedAsync += HandleDisconnectionAsync;
```

Comment:
since we've already set the disconnection handler and incoming message handler in the constructor (MqttTransportHandler), do we need to set them again here?

Reply:
I'll make it so we += after open and -= before closing. That way we don't have += at constructor time and += at open time